### PR TITLE
Feat/cap 280 improve user instruction info

### PR DESCRIPTION
**Pull request by pecodez**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
This PR adds a property to dockerfilePackages for the installCommand used within a RUN instruction. This allows for a more targetted output and solves the issue of long instructions with arbitrary shell commands and sometimes sensitive information being output for each package that introduces a vuln.

#### What are the relevant tickets?
https://snyksec.atlassian.net/browse/CAP-280

### CODE REVIEW

File: test/lib/instructions-parser.spec.ts [hunk 1]
```diff
@@ -60,6 +60,16 @@ describe("getPackagesFromRunInstructions", () => {
         );
       },
     );
+    test("given an image built with --build-args, expect ARG values to be stripped from RUN instruction", () => {
```
Comment (agatakrajewska) [inline test/lib/instructions-parser.spec.ts:63]: Tiny: I think the flag is actually `--build-arg` without the s

**Comment by agatakrajewska** [comment]
Great work and also amazing test improvements! 🥇

File: test/lib/instructions-parser.spec.ts [hunk 2]
```diff
@@ -22,17 +31,71 @@ describe("instruction parser", () => {
...
+      const filePath = resolve(
+        "test/fixtures/dockerfiles",
+        fixture,
+        "Dockerfile",
+      );
+      const content = readFileSync(filePath, "utf-8").toString();
+      const dockerFile = DockerfileParser.parse(content);
+      const results = getPackagesFromDockerfile(dockerFile);
+      Object.keys(results).forEach((pkgName) =>
+        expect(results[pkgName].installCommand).toBeDockerPackageInstallCommand(
```
Comment (agatakrajewska) [inline test/lib/instructions-parser.spec.ts:95]: Nice!

File: test/lib/instructions-parser.spec.ts (hunk 1 above)
Comment (pecodez) [reply test/lib/instructions-parser.spec.ts:63]: Good spot, thanks

File: test/lib/dockerfile.spec.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    ],
+  ];
+  // tslint:disable-next-line: variable-name
+  test.each<TestCaseTuple>(cases)("%s", async (_description, item) => {
+    const { fixture, expected } = item;
+    const pathToDockerFile = getDockerfileFixture(fixture);
+    const actual = await dockerFile.readDockerfileAndAnalyse(pathToDockerFile);
+    expect(actual.baseImage).toEqual(expected.baseImage);
+    expect(actual.error).toEqual(expected.error);
+    Object.keys(expected.dockerfilePackages).forEach((pkg) => {
```
Comment (shaninja) [inline test/lib/dockerfile.spec.ts:253]: 🙀  looping expects inside `.each` -- how does the failure message look when something fails, and is it easy enough to reason with and understand the issue?

**Review by shaninja — APPROVED** [review]
Thank you for all the jest conversions! 🙇 👑 

and you got some serious jest-fu! 🥷

File: lib/dockerfile/instruction-parser.ts [hunk 4]
```diff
@@ -78,12 +86,25 @@ function getPackagesFromRunInstructions(runInstructions: string[]) {
   }, {});
 }
 
-function removeRunDefFromInstruction(instruction: string) {
+/**
+ * Return the instruction text without any of the image prefixes
+ * @param instruction the full RUN instruction extracted from image
+ * @returns string
```
[the remark this answers is not in the document]
Comment (pecodez) [reply lib/dockerfile/instruction-parser.ts:92]: It's a good point. I've been using JSDOC for years and old habits die hard but you're right; keeping JSDOC in sync with TS can be a pain!

File: lib/dockerfile/instruction-parser.ts [hunk 5]
```diff
@@ -78,12 +86,25 @@ function getPackagesFromRunInstructions(runInstructions: string[]) {
...
   let cleanedInstruction = instruction;
-  const runDefs = ["RUN ", "/bin/sh ", "RUN /bin/sh"];
+  const runDefs = ["RUN ", "/bin/sh ", "-c "];
+  const argsPrefixRegex = /^\|\d .*?=/;
+
   for (const runDef of runDefs) {
     if (cleanedInstruction.startsWith(runDef)) {
       cleanedInstruction = cleanedInstruction.slice(runDef.length);
+      if (runDef === runDefs[0] && argsPrefixRegex.test(cleanedInstruction)) {
+        const match = installRegex.exec(cleanedInstruction);
```
[the remark this answers is not in the document]
Comment (pecodez) [reply lib/dockerfile/instruction-parser.ts:103]: Using the `match.index` property of the `result` on #105 to get the start char once a match has been found.

File: test/matchers/dockerPackageInstallCommand.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  }
+}
+
+export function toBeDockerPackageInstallCommand(received, pkgName) {
+  const installCmdRegex = /^(rpm\s+-i|rpm\s+--install|apk\s+((--update|-u|--no-cache)\s+)*add(\s+(--update|-u|--no-cache))*|apt-get\s+((--assume-yes|--yes|-y)\s+)*install(\s+(--assume-yes|--yes|-y))*|apt\s+((--assume-yes|--yes|-y)\s+)*install|yum\s+install|aptitude\s+install)\s+/;
+  const pass =
+    (installCmdRegex.test(received) &&
+      received.indexOf(pkgName) > -1 &&
+      !RegExp("/ +|\t|\n/", "g").test(received) &&
```
[the remark this answers is not in the document]
Comment (pecodez) [reply test/matchers/dockerPackageInstallCommand.ts:14]: Good catch, that is supposed to match multiple spaces so should be ` {2,}` not ` +`. I will quickly revisit this 👍🏻

File: test/matchers/dockerPackageInstallCommand.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+declare global {
+  namespace jest {
+    interface Matchers<R> {
+      toBeDockerPackageInstallCommand(pkgName: string): R;
+    }
+  }
+}
+
+export function toBeDockerPackageInstallCommand(received, pkgName) {
+  const installCmdRegex = /^(rpm\s+-i|rpm\s+--install|apk\s+((--update|-u|--no-cache)\s+)*add(\s+(--update|-u|--no-cache))*|apt-get\s+((--assume-yes|--yes|-y)\s+)*install(\s+(--assume-yes|--yes|-y))*|apt\s+((--assume-yes|--yes|-y)\s+)*install|yum\s+install|aptitude\s+install)\s+/;
```
[the remark this answers is not in the document]
Comment (pecodez) [reply test/matchers/dockerPackageInstallCommand.ts:10]: Agreed!

**Review by shaninja — APPROVED** [review]

**Pull request merged**